lib/trino-filesystem/src/main/java/io/trino/filesystem/UnrecoverableIOException.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem;
+
+import com.google.common.base.Throwables;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Optional;
+
+/**
+ * This exception is for stopping retries for FileSystem calls that shouldn't be retried.
+
+ * Use this exception for cases, where explicit message that IO has failed and retrying doesn't make any sense in context of operation success.
+ * There could be various reasons such as 404, 403, 401 related error codes.
+
+ * By default, all third party (AWS, Azure, GCP) SDKs will retry appropriate exceptions (either client side IO errors, or 500/503), so there is
+ * no need to retry those additionally.
+ * If any custom retry behavior is needed, it is advised to change SDK's retry handlers, rather than introducing outer retry loop, which combined
+ * with SDKs default retry, could lead to prolonged, unnecessary retries
```

Comment:
I might be wrong here, but I think the exception description should explain what to do when catching the exception, not when to throw it.

The description explaining different kind of errors that this can wrap should be moved to the constructor.

```suggestion
 * Trino file system exception. When the cause is not fatal, the internal implementation should have already attempted retries, so no additional retries should be done when this is caught.
```

Reply:
I'll leave your comment in exception's javadoc, I'll move more verbose explanation on top of new method `TrinoFileSystem#isUnrecoverableException`